lib/puppet/provider/gpg/gpg.rb

```diff
@@ -0,0 +1,318 @@
+            '--export-secret-subkeys', "#{sub_gpg_id}!",
+          ]
+          exported_subkey = gpg_exec_options(*gpg_args, options: exec_options)
+          if exported_subkey.empty?
+            raise Puppet::ExecutionFailure, 'Failed to export GPG subkey'
+          end
+        end
+
+        # Import subkey into a host-specific keyring
+        prep_new_gpg(build_sub_gpg)
+        exec_options = { :stdin_string => exported_subkey }
+        gpg_exec_options('--homedir', build_sub_gpg, '--import', options: exec_options)
+      end
+    end
+
+    # Generate a host-specific RPM
+    # (Containing either the primary keyring or the host-specific keyring
+    # depending on the value of :subkey)
+    build_gpg = (resource[:subkey] ? build_sub_gpg : build_pri_gpg)
+    rpm_args = [
+      '--name', rpm_base_filename,
+      '--version', ((existing_rpms.map { |r| version_from_rpm_filename(r) }.max || 0) + 1),
+      '--summary', "#{resource[:name]} for #{resource[:hostname]}",
+      '--description', "#{resource[:name]} for #{resource[:hostname]}",
+      gpg_rpm_filespec(build_gpg, [], '0700'),
+      gpg_rpm_filespec(build_gpg, ['gpg.conf'], '0600'),
+      gpg_rpm_filespec(build_gpg, ['pubring.gpg'], '0600'),
+      # Include public GPG key file from primary keyring directory
+      gpg_rpm_filespec(build_pri_gpg, [resource[:deploy_pub_file]], '0600'),
+      # Other keyring files should not be required, but include them anyway
+      # (Explicitly listing the required files in addition to this wildcard
+      # ensures that a failure will occur if any of the required files are
+      # missing due to an undetected error during generation)
+      gpg_rpm_filespec(build_gpg, ['*'], '0600'),
+    ]
+    if File.exist?(File.join(build_gpg, 'secring.gpg'))
+      rpm_args << [
+        gpg_rpm_filespec(build_gpg, ['secring.gpg'], '0600'),
+      ]
+    end
+    if File.exist?(File.join(build_gpg, 'private-keys-v1.d'))
+      rpm_args << [
+        gpg_rpm_filespec(build_gpg, ['private-keys-v1.d'], '0700'),
+        gpg_rpm_filespec(build_gpg, ['private-keys-v1.d', '*'], '0600'),
+      ]
+    end
+    Dir.chdir(rpm_path) do
+      katello_certs_gen_rpm(*rpm_args)
+    end
+  end
+
+  def deploy?
+    return false unless resource[:deploy]
+    return true if resource[:regenerate_ca]
+    return true if resource[:subkey] and resource[:regenerate]
+    return true unless gpg_exists?(resource[:deploy_gpg])
+    pub_file = File.join(resource[:deploy_gpg], resource[:deploy_pub_file])
+    return true unless File.exist?(pub_file)
+    rpm_file = latest_rpm
+    if rpm_file
+      !system("rpm --verify -p #{rpm_file} &>/dev/null")
+    else
+      `yum check-update #{rpm_base_filename} &>/dev/null`
+      $?.exitstatus == 100
+    end
+  end
+
+  def deploy!
+    if(system("rpm -q #{rpm_base_filename} &>/dev/null"))
+      rpm('-e', rpm_base_filename)
+    end
+    rpm_file = latest_rpm
+    if rpm_file
+      rpm('-Uvh', '--force', rpm_file)
```

Comment:
Do we really need `--force` ?

Reply:
Not sure ... I copied that code from https://github.com/theforeman/puppet-certs/blob/master/lib/puppet/provider/katello_ssl_tool.rb#L77